cmd/oras/root/restore.go

```diff
@@ -0,0 +1,246 @@
+
+Example - Restore all tagged artifacts:
+  oras restore --input hello localhost:5000/hello
+
+Example - Exclude referrers when restoring artifacts:
+  oras restore --input hello --exclude-referrers localhost:5000/hello
+
+Example - Use Referrers API for discovering referrers:
+  oras restore --input hello --distribution-spec v1.1-referrers-api localhost:5000/hello
+
+Example - Use Referrers Tag Schema for discovering referrers:
+  oras restore --input hello --distribution-spec v1.1-referrers-tag localhost:5000/hello
+
+Example - Restore to an insecure registry:
+  oras restore --input hello --insecure localhost:5000/hello:v1
+
+Example - Restore to a plain HTTP registry (no TLS):
+  oras restore --input hello --plain-http localhost:5000/hello:v1
+
+Example - Set custom concurrency level:
+  oras restore --input hello --concurrency 6 localhost:5000/hello:v1
+`,
+		Args: oerrors.CheckArgs(argument.Exactly(1), "the targets to restore to"),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if err := option.Parse(cmd, &opts); err != nil {
+				return err
+			}
+
+			// parse repo and tags
+			var err error
+			opts.repository, opts.tags, err = parseArtifactReferences(args[0])
+			if err != nil {
+				return err
+			}
+
+			opts.DisableTTY(opts.Debug, false)
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.Printer.Verbose = true // always print verbose output
+			return runRestore(cmd, &opts)
+		},
+	}
+
+	// required flag
+	cmd.Flags().StringVar(&opts.input, "input", "", "path to the OCI layout, either a tar archive (*.tar) or a directory")
+	_ = cmd.MarkFlagRequired("input")
+	// optional flags
+	cmd.Flags().BoolVar(&opts.excludeReferrers, "exclude-referrers", false, "restore artifacts excluding their referrers")
+	cmd.Flags().BoolVar(&opts.dryRun, "dry-run", false, "simulate the restore process without actually uploading any artifacts")
+	cmd.Flags().IntVarP(&opts.concurrency, "concurrency", "", 3, "concurrency level")
+	opts.EnableDistributionSpecFlag()
+	// apply flags
+	option.ApplyFlags(&opts, cmd.Flags())
+	return oerrors.Command(cmd, &opts.Remote)
+}
+
+func runRestore(cmd *cobra.Command, opts *restoreOptions) error {
+	if opts.input == "" {
+		return errors.New("the input path cannot be empty")
+	}
+	if _, err := os.Stat(opts.input); err != nil {
+		return fmt.Errorf("failed to access input path %q: %w", opts.input, err)
+	}
+	startTime := time.Now() // start timing the restore process
+	ctx, logger := command.GetLogger(cmd, &opts.Common)
+
+	// prepare the target registry
+	dstRepo, err := opts.NewRepository(opts.repository, opts.Common, logger)
+	if err != nil {
+		return fmt.Errorf("failed to prepare target repository %q: %w", opts.repository, err)
+	}
+	statusHandler, metadataHandler := display.NewRestoreHandler(opts.Printer, opts.TTY, dstRepo)
+
+	// prepare the source OCI store
+	var srcOCI oras.ReadOnlyGraphTarget
+	if strings.HasSuffix(opts.input, ".tar") {
```

Comment:
You may need to `stat` `opts.input` first to check if it is a regular file or not. Unlike `oras backup`, `oras restore` should try its best to determine the input type.

Here's the pseudocode:
1. `stat` the input if it is a folder
2. If it is a folder, use it as a folder in OCI image layout.
3. else if it is a regular file, check it's filename.
   1. if it ends with case-insensitive `.tar`, use it as a tar file in OCI image layout.
   2. else if it has magic number `ustar` at offset 257, use it as a tar file in OCI image layout.
   3. else, report error: unsupported archive type.
4. else, report error: not a regular file or directory.